FirebaseRemoteConfig/Sources/RCNConfigRealtime.m

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "FirebaseRemoteConfig/Sources/RCNConfigRealtime.h"
+#import <Foundation/Foundation.h>
+#import "FirebaseRemoteConfig/Sources/RCNConfigConstants.h"
+
+@implementation FIRConfigUpdateListenerRegistration {
+  RCNConfigRealtime *_realtimeClient;
+  FIRConfigUpdateCompletion completionHandler;
+}
+
+- (instancetype)initWithClient:(RCNConfigRealtime *)realtimeClient
+             completionHandler:(FIRConfigUpdateCompletion)completionHandler {
+  self = [super init];
+  if (self) {
+    _realtimeClient = realtimeClient;
+    completionHandler = completionHandler;
+  }
+  return self;
+}
+
+- (void)remove {
+  [self->_realtimeClient removeConfigUpdateListener:completionHandler];
+}
+
+@end
+
+@interface RCNConfigRealtime ()
+
+@property(strong, atomic, nonnull) NSMutableSet<FIRConfigUpdateCompletion> *listeners;
+@property(strong, atomic, nonnull) dispatch_queue_t realtimeLockQueue;
+
+@end
+
+@implementation RCNConfigRealtime
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _listeners = [NSMutableSet alloc];
+    _realtimeLockQueue = [RCNConfigRealtime realtimeRemoteConfigSerialQueue];
+  }
+
+  return self;
+}
+
+/// Singleton instance of serial queue for queuing all incoming RC calls.
++ (dispatch_queue_t)realtimeRemoteConfigSerialQueue {
+  static dispatch_once_t onceToken;
+  static dispatch_queue_t realtimeRemoteConfigQueue;
+  dispatch_once(&onceToken, ^{
+    realtimeRemoteConfigQueue =
+        dispatch_queue_create(RCNRemoteConfigQueueLabel, DISPATCH_QUEUE_SERIAL);
+  });
+  return realtimeRemoteConfigQueue;
+}
+
+- (void)beginRealtimeStream {
+}
+
+- (void)pauseRealtimeStream {
+}
+
+- (FIRConfigUpdateListenerRegistration *)addConfigUpdateListener:
+    (FIRConfigUpdateCompletion)completionHandler {
+  __weak RCNConfigRealtime *weakSelf = self;
+  dispatch_async(_realtimeLockQueue, ^{
+    __strong RCNConfigRealtime *strongSelf = weakSelf;
+    [strongSelf->_listeners addObject:completionHandler];
+    [strongSelf beginRealtimeStream];
```

Comment:
follow up to the above, if we allow multiple listeners then we would only want to call `beginRealtimeStream` for the first one? (similar to how you handle `pauseRealtimeStream`)

Reply:
Yeah we would only want to call beginRealtimeStream when there is at least 1 listener present, there is no current stream connections, and the app is in foreground. In the next PRs I was going to have a helper function that checks this and place it in the beginRealtimeStream method. I want to place the check there because beginRealtimeStream will also be called by retry functions which will also need to have this check.